Add access log handler for HTTP stream filter.

qiwzhang opened the pull request:
For https://github.com/lyft/envoy/issues/393

mattklein123 commented:
this looks reasonable to me. tests please.

mattklein123 commented on include/envoy/http/filter.h:
```diff
@@ -268,6 +268,22 @@ class StreamFilter : public StreamDecoderFilter, public StreamEncoderFilter {};
+   * Called with request_info data.
+   * @param request_info: the request_info data.
+   */
+  virtual void logRequestInfo(const Http::AccessLog::RequestInfo& request_info) PURE;
```
can we also pass request/response headers (optionally) like the other access log calls? Might be useful for others in the future.

qiwzhang commented:
@RomanDzhabarov and @mattklein123 
could you help to review it?

qiwzhang commented:
@RomanDzhabarov and @mattklein123.  I changed to use Http::Accesslog::Instance directly and added a test case.  Could you help to review it again?  Thanks.

mattklein123 commented on include/envoy/http/filter.h:
```diff
@@ -292,6 +292,12 @@ class FilterChainFactoryCallbacks {
   virtual void addStreamFilter(Http::StreamFilterPtr filter) PURE;
+
+  /**
+   * Add an access log handler that is called after response is send.
```
"Add an access log handler that is called when the stream is destroyed." (A response may not have been sent)

qiwzhang replied:
Done
